stream_alert/alert_processor/outputs/output_base.py

```diff
@@ -116,62 +116,38 @@ def get_all_outputs(cls):
-            and locally on disk in tmp
-        get_config_service: the name of the service used by the config to store any
-            configured outputs for this service. implemented by some subclasses, but
-            subclass is not required to implement
-        format_output_config: returns a formatted version of the outputs configuration
-            that is to be written to disk
-        get_user_defined_properties: returns any properties for this output that must be
-            provided by the user. must be implemented by subclasses
-        dispatch: handles the actual sending of alerts to the configured service. must
-            be implemented by subclass
-    """
-    __metaclass__ = ABCMeta
-    __service__ = NotImplemented
+class OutputCredentialsProvider(object):
+    """OutputCredentialsProvider is a helper service to OutputDispatcher that helps it load
+       credentials that are housed on AWS S3, or cached locally.
 
-    # How many times it will attempt to retry something failing using backoff
-    MAX_RETRY_ATTEMPTS = 5
-
-    # _DEFAULT_REQUEST_TIMEOUT indicates how long the requests library will wait before timing
-    # out for both get and post requests. This applies to both connection and read timeouts
-    _DEFAULT_REQUEST_TIMEOUT = 3.05
-
-    def __init__(self, config):
-        self.account_id = os.environ['AWS_ACCOUNT_ID']
-        self.region = REGION
-        self.secrets_bucket = '{}.streamalert.secrets'.format(os.environ['STREAMALERT_PREFIX'])
-        self.config = config
-
-    @staticmethod
-    def _local_temp_dir():
-        """Get the local tmp directory for caching the encrypted service credentials
+       OutputDispatcher implementations may require credentials to authenticate with an external
+       gateway. All credentials for OutputDispatchers are to be stored in a single bucket on AWS S3
+       and are encrypted with AWS KMS. When the OutputDispatchers are booted, this these encrypted
+       credentials are downloaded and cached locally on the filesystem. Then, AWS KMS is used to
+       decrypt the credentials when in use.
 
-        Returns:
-            str: local path for stream_alert_secrets tmp directory
-        """
-        temp_dir = os.path.join(tempfile.gettempdir(), "stream_alert_secrets")
+    Public methods:
+        load_credentials: Returns a dict of the credentials requested
+        get_local_credentials_temp_dir(): Returns full path to a temporary directory where all
+            encrypted credentials are cached.
 
-        # Check if this item exists as a file, and remove it if it does
-        if os.path.isfile(temp_dir):
-            os.remove(temp_dir)
+    """
 
-        # Create the folder on disk to store the credentials temporarily
-        if not os.path.exists(temp_dir):
-            os.makedirs(temp_dir)
+    def __init__(self, config, defaults, service_name):
+        self._config = config
+        self._region = REGION
```

Comment:
this might be silly - but since `REGION` is imported as a global, it doesn't really need to be set as a local/instance property. not a huge deal to leave it though

Reply:
I'll leave as-is; I'm a big proponent of classes having local references of everything they need to operate; a very *Dependency-Injection-centric* style.